### ISSUE TITLE
Unexpected mapping with BSS.Align.matchAtoms

### ISSUE THREAD

**Issue by jmichel80** [michellab/BioSimSpace#76]
In the following example that can be reproduced with the attached input files 

[31to32.zip](https://github.com/michellab/BioSimSpace/files/3038989/31to32.zip)

The intuitive result from executing prepareFEP.ipynb in this[ commit](https://github.com/michellab/BioSimSpace/commit/7411e23cfb87ee6424aa946c3a87b3bb12efb4a8) using as input options

![image](https://user-images.githubusercontent.com/11722212/55483554-b825bf80-561e-11e9-8e1a-5b474865096d.png)

and the following arguments to matchAtoms

`mappings, scores = BSS.Align.matchAtoms(lig1, lig2, matches=10, prematch=prematch, return_scores=True, scoring_function="RMSDalign", timeout=10*BSS.Units.Time.second)
`

is to return a mapping where 

Atom C18  (index 18) in 31vac.parm7 is mapped to atom C32 (index 32) in 32vac.parm7
and
Atom C20 (index 19) in 31vac.parm7 is mapped to atom C17  (index 31) in 32vac.parm7

![image](https://user-images.githubusercontent.com/11722212/55483632-e2777d00-561e-11e9-80d8-49754252d103.png)

Instead none of the mappings returned have the desired mapping. All solutions include

C18 in 31vac.parm7  (index 17) mapped to C17 in 32vac.parm7 (index 16)
C20 in 31vac.parm7 (index 19) mapped to C32 in 32vac.parm7 (index 31)

As a result the fused pyridine ring is attached to a different 'face' of the phenyl moiety which leads to large scale rearrangements in the context of the binding site and poor convergence of calculated free energy changes. 

![image](https://user-images.githubusercontent.com/11722212/55483819-4d28b880-561f-11e9-8552-285af3f8a3e7.png)

The same behaviour can be reproduced by passing directly as mapping argument the file ''31to32.map'' which uses the same mappings as the top mapping produced by BSS.Align.matchAtoms

I then sought to obtain the desired behaviour by passing as optional argument to mapping the file ''31to32-custom.map''

which swaps the erroneous mappings

```
julien@ubuntu:/mnt/vmshared/Cresset/mcl1_4julien/31to32$ diff 31to32.map 31to32-custom.map 
2c2
< 17, 16 
---
> 17, 31 
4c4
< 19, 31 
---
> 19, 16 

```

However this leads to the following error

```
---------------------------------------------------------------------------
IncompatibleError                         Traceback (most recent call last)
<ipython-input-22-6cb20ac7a6ce> in <module>
      4 lig1 = BSS.Align.rmsdAlign(lig1, lig2, mapping)
      5 # Merge the two ligands based on the mapping.
----> 6 merged = BSS.Align.merge(lig1, lig2, mapping)
      7 # Create a composite system
      8 system1.removeMolecules(lig1)

~/sire.app/lib/python3.7/site-packages/BioSimSpace-2018.1.1+842.g7ed7835.dirty-py3.7.egg/BioSimSpace/Align/_align.py in merge(molecule0, molecule1, mapping, property_map0, property_map1)
    440 
    441     # Create and return the merged molecule.
--> 442     return molecule0._merge(molecule1, mapping, property_map0=property_map0, property_map1=property_map1)
    443 
    444 def _score_rmsd(molecule0, molecule1, mappings, is_align=False):

~/sire.app/lib/python3.7/site-packages/BioSimSpace-2018.1.1+842.g7ed7835.dirty-py3.7.egg/BioSimSpace/_SireWrappers/_molecule.py in _merge(self, other, mapping, property_map0, property_map1)
   2257                 if c1.connectionType(_SireMol.AtomIdx(x), _SireMol.AtomIdx(y)) != \
   2258                    conn.connectionType(idx, idy):
-> 2259                        raise _IncompatibleError("Merge has changed the molecular connectivity! "
   2260                                                 "Check your atom mapping.")
   2261 

IncompatibleError: Merge has changed the molecular connectivity! Check your atom mapping.
```

@lohedges 

- any idea why the mapping code fails to align C18 to C17 ? Increasing the timeout to 60 seconds had no effect. 

- why the custom mapping defined in 31to32-custom.map is invalid ?

**Comment by jmichel80** [comment]
An update, the reason 31to32-custom.map failed is that it is necessary to swap all C2 symmetry equivalent atoms along the C16-C19 axis in 31vac in the mapping file in order to maintain the original topology. 

The attached mapping file generates the desired output

[31to32-custom-v2.zip](https://github.com/michellab/BioSimSpace/files/3039078/31to32-custom-v2.zip)

![image](https://user-images.githubusercontent.com/11722212/55485841-28cedb00-5623-11e9-9a65-8153ec92f8a8.png)

**Comment by lohedges** [comment]
Is the symmetry swapping something that should be handled by the MCS matcher? if so, we'll need to work out a way of incorporating this logic into the Sire findMCS routine.

**Comment by jmichel80** [comment]
Yes the idea of RMSDAlign is to return different MCS that include topologically equivalent mappings, and then score the various solutions by a spatial metric. At least that's how it was done with FESetup.

You seem to imply that the MCS matcher would not for instance find two different ways of mapping a R-C6H5 ring onto a R-C6H4F ring which is not  my understanding of how the algorithm works.

**Comment by lohedges** [comment]
Yes, I totally agree with what you're saying, that is my understanding of the MCS matching algorithm too. What I'm saying is that BioSimSpace works with whatever the SIre findMCS returns, which doesn't seem to include the match that you want.

Does FESetup use Sire for MCS matching? (In my head I'm thinking that it uses RDKit.)

One possible issue is that the findMCSmatches only gives the set of matches with the same number of atoms. (Once the number of atoms matched increases, the previous matches are discarded.) Is it possible that the match you want (with best alignment) actually involves fewer atoms in the mapping?
Also, the RMSDAlign only considers RMSD between atoms that are part of the mapping, not the whole ligands (you need a consistent number of atoms for the RMSD in SIre) so perhaps this is affecting the scoring.

**Comment by skfegan** [comment]
FESetup uses findmcs from rdkit.

**Comment by jmichel80** [comment]
The desired mapping has the same MCS size than those  returned currently. It would have a lower RMSD  value than those currently returned since some atoms in the MCS are spatially closer after alignment.

The issue seems to be that findMCS fails to find the mapping in the first instance.

**Comment by lohedges** [comment]
Thanks, I'll have a look at the Sire code to see if I can figure out what it's missing. If anyone could think of a minimal test case that currently fails, that would be great.

**Comment by lohedges** [comment]
For reference, the RDKit prematch that is now implemented in BioSImSpace maps 17 to 16, whereas 19 is now mapped to 19. Since FESetup uses RDKit for its MCS matcher, I'd be interested to see if it successfully generates your custom map.

**Comment by lohedges** [comment]
I've now implemented a spatially closest ranking of the RDKit prematch, following the implementation in FESetup. This correctly returns the mapping that you desire without the need for a custom mapping. You just need to pass `use_rdkit_prematch=True` to the `matchAtoms` function.

**Comment by lohedges** [comment]
I was getting frustrated at how unreliable our MCS matching is, so have taken time to re-write the `BioSimSpace.Align.matchAtoms` function so that it exclusively uses RDKit. I had to make some slight adjustments to the FESetup approach (mostly due to API changes in RDKit and for consistency with the existing BioSimSpace behaviour) but we now have a much faster and more robust atom mapping code. This means that we no-longer use the Sire `findMCS` functionality and don't need to do multiple combinations of matches, i.e. with/without heavy atoms, etc.

I've tested all of the problematic examples that you've posted in the issue tracker and can recover the desired mappings in all cases.
